File: apps/augury-signal/tests/test_lmsr.py
"""Tests for the reference LMSR implementation.

The invariants asserted here are the same ones the C++ engine's Catch2 suite
checks against the shared golden vectors.
"""

from __future__ import annotations

import math

import pytest

from augury_signal.engine.lmsr import (
    LMSRMarket,
    calibrate_b_from_depth,
    cost,
    price_yes,
    prices,
    shares_to_move_price,
    trade_cost,
    worst_case_subsidy,
)


class TestCostFunction:
    def test_uniform_state_is_b_ln_k(self):
        """C(0) = b*ln(K) for a K-outcome market."""
        assert cost([0.0, 0.0], 100.0) == pytest.approx(100.0 * math.log(2))
        assert cost([0.0] * 4, 50.0) == pytest.approx(50.0 * math.log(4))

    def test_does_not_overflow_at_realistic_scale(self):
        """q/b of 500 overflows exp() in float64; log-sum-exp must not.

        The naive formula returns inf here, and every price derived from it
        becomes nan.
        """
        value = cost([50_000.0, 0.0], 100.0)
        assert math.isfinite(value)
        # C(q) -> max(q) as the leading term dominates.
        assert value == pytest.approx(50_000.0, rel=1e-9)

    def test_translation_invariance(self):
        """Adding a constant to every q raises C by exactly that constant."""
        base = cost([10.0, -5.0], 25.0)
        assert cost([110.0, 95.0], 25.0) == pytest.approx(base + 100.0)

    def test_rejects_bad_inputs(self):
        with pytest.raises(ValueError):
            cost([1.0], 10.0)
        with pytest.raises(ValueError):
            cost([1.0, 2.0], 0.0)
        with pytest.raises(ValueError):
            cost([1.0, math.inf], 10.0)


class TestPrices:
    def test_prices_form_a_distribution(self):
        p = prices([120.0, -30.0, 45.0], 40.0)
        assert math.fsum(p) == pytest.approx(1.0)
        assert all(0.0 < x < 1.0 for x in p)

    def test_symmetric_state_is_even(self):
        assert prices([0.0, 0.0], 100.0) == pytest.approx([0.5, 0.5])

    def test_binary_helper_agrees_with_vector_form(self):
        vector = prices([37.0, -12.0], 55.0)[0]
        assert price_yes(37.0, -12.0, 55.0) == pytest.approx(vector)

    def test_logit_is_linear_in_share_imbalance(self):
        """logit(p_yes) = (q_yes - q_no)/b — the identity everything else uses."""
        b = 80.0
        for q_yes, q_no in [(0.0, 0.0), (100.0, 20.0), (-60.0, 15.0)]:
            p = price_yes(q_yes, q_no, b)
            assert math.log(p / (1 - p)) == pytest.approx((q_yes - q_no) / b)


class TestTrades:
    def test_zero_trade_is_free(self):
        assert trade_cost([10.0, 4.0], [0.0, 0.0], 30.0) == pytest.approx(0.0)

    def test_buying_raises_the_price(self):
        market = LMSRMarket(b=100.0)
        before = market.price
        market.buy_yes(50.0)
        assert market.price > before

    def test_round_trip_is_free_up_to_float_error(self):
        """Buy then sell the same size returns to the original state."""
        market = LMSRMarket(b=100.0)
        paid = market.buy_yes(75.0)
        refund = market.buy_yes(-75.0)
        assert paid + refund == pytest.approx(0.0, abs=1e-9)
        assert market.price == pytest.approx(0.5)

    def test_cost_bounded_by_shares(self):
        """A YES share can never cost more than 1: it pays at most 1 on resolution."""
        market = LMSRMarket(b=100.0)
        shares = 250.0
        assert market.buy_yes(shares) < shares

    def test_shares_to_move_price_is_exact(self):
        market = LMSRMarket(b=140.0)
        needed = shares_to_move_price(market.price, 0.8, market.b)
        market.buy_yes(needed)
        assert market.price == pytest.approx(0.8, abs=1e-12)

    def test_move_to_price_hits_target(self):
        market = LMSRMarket(b=60.0, q_yes=30.0)
        for target in (0.1, 0.5, 0.97):
            market.move_to_price(target)
            assert market.price == pytest.approx(target, abs=1e-9)

    def test_worst_case_subsidy(self):
        assert worst_case_subsidy(2, 100.0) == pytest.approx(100.0 * math.log(2))


class TestRecalibration:
    def test_price_is_preserved(self):
        """Changing b is a liquidity update, not new information about the outcome.

        If it moved the price, the simulation would be injecting signal that no
        trader ever expressed.
        """
        market = LMSRMarket(b=100.0)
        market.move_to_price(0.73)
        before = market.price
        market.recalibrate(925.0)
        assert market.price == pytest.approx(before, abs=1e-12)
        assert market.b == 925.0

    def test_higher_b_means_less_price_impact(self):
        thin, thick = LMSRMarket(b=10.0), LMSRMarket(b=10_000.0)
        thin.buy_yes(100.0)
        thick.buy_yes(100.0)
        assert abs(thin.price - 0.5) > abs(thick.price - 0.5)

    def test_rejects_non_positive(self):
        with pytest.raises(ValueError):
            LMSRMarket(b=100.0).recalibrate(0.0)


class TestDepthCalibration:
    def test_matches_the_defining_equation(self):
        """b = depth / (logit(ask) - logit(bid))."""
        bid, ask, depth = 0.30, 0.35, 500.0
        width = math.log(ask / (1 - ask)) - math.log(bid / (1 - bid))
        got = calibrate_b_from_depth(depth, depth, bid, ask)
        assert got == pytest.approx(depth / width, rel=1e-9)

    def test_calibrated_b_reproduces_the_observed_move(self):
        """The whole point: consuming `depth` shares walks bid -> ask."""
        bid, ask, depth = 0.20, 0.26, 800.0
        b = calibrate_b_from_depth(depth, depth, bid, ask)
        market = LMSRMarket(b=b)
        market.move_to_price(bid)
        market.buy_yes(depth)
        assert market.price == pytest.approx(ask, abs=1e-9)

    def test_deeper_book_gives_larger_b(self):
        thin = calibrate_b_from_depth(10.0, 10.0, 0.4, 0.45)
        thick = calibrate_b_from_depth(10_000.0, 10_000.0, 0.4, 0.45)
        assert thick > thin

    def test_side_selection_uses_the_named_depth(self):
        """Real books are lopsided, so 'bid' and 'ask' genuinely differ."""
        bid_only = calibrate_b_from_depth(40.0, 69_298.0, 0.03, 0.04, side="bid")
        ask_only = calibrate_b_from_depth(40.0, 69_298.0, 0.03, 0.04, side="ask")
        both = calibrate_b_from_depth(40.0, 69_298.0, 0.03, 0.04, side="mean")
        assert bid_only < both < ask_only

    @pytest.mark.parametrize(
        "depth_bid,depth_ask,bid,ask",
        [
            (100.0, 100.0, 0.6, 0.4),   # crossed
            (100.0, 100.0, 0.5, 0.5),   # locked
            (None, None, 0.3, 0.4),     # no depth
            (0.0, 0.0, 0.3, 0.4),       # zero depth
            (100.0, 100.0, None, 0.4),  # one-sided quote
        ],
    )
    def test_unusable_books_fall_back(self, depth_bid, depth_ask, bid, ask):
        """Thin prediction markets produce these constantly; they must not raise."""
        assert calibrate_b_from_depth(depth_bid, depth_ask, bid, ask, fallback=42.0) == 42.0

    def test_result_is_clamped(self):
        huge = calibrate_b_from_depth(1e12, 1e12, 0.499, 0.501)
        assert huge <= 1.0e6


class TestSignalStep:
    def test_full_responsiveness_snaps_to_target(self):
        market = LMSRMarket(b=100.0)
        market.step_from_signal(0.8, responsiveness=1.0)
        assert market.price == pytest.approx(0.8, abs=1e-9)

    def test_partial_responsiveness_undershoots(self):
        """Below 1.0 the market integrates the signal instead of restating it."""
        market = LMSRMarket(b=100.0)
        market.step_from_signal(0.9, responsiveness=0.25)
        assert 0.5 < market.price < 0.9

    def test_repeated_steps_converge(self):
        market = LMSRMarket(b=100.0)
        for _ in range(100):
            market.step_from_signal(0.7, responsiveness=0.3)
        assert market.price == pytest.approx(0.7, abs=1e-6)

    def test_max_shares_caps_a_single_step(self):
        market = LMSRMarket(b=100.0)
        shares, _ = market.step_from_signal(0.999, responsiveness=1.0, max_shares=10.0)
        assert shares == pytest.approx(10.0)

    def test_rejects_bad_responsiveness(self):
        with pytest.raises(ValueError):
            LMSRMarket(b=100.0).step_from_signal(0.6, responsiveness=1.5)

File: apps/augury-engine/tests/test_lmsr.cpp
// Correctness gate for the C++ engine.
//
// The golden-vector cases here are the point of this file: they assert that
// this implementation reproduces the Python reference in
// `apps/augury-signal/augury_signal/engine/lmsr.py` to 1e-9 on a fixed set of
// inputs. Two independent implementations of the same equations agreeing to
// nine digits is meaningful evidence that both are right; either one agreeing
// with itself is not.
//
// If these fail after an intentional change to the Python side, regenerate with
// `python -m augury_signal.golden` and re-read the diff before accepting it.

#include <catch2/catch_approx.hpp>
#include <catch2/catch_test_macros.hpp>
#include <nlohmann/json.hpp>

#include <cmath>
#include <filesystem>
#include <fstream>
#include <optional>
#include <vector>

#include "augury/backtest.hpp"
#include "augury/lmsr.hpp"

using augury::Cost;
using augury::LmsrMarket;
using augury::PriceYes;
using Catch::Approx;

namespace {

constexpr double kTolerance = 1e-9;

/// Locate `schemas/testdata/golden_vectors.json` by walking up from the build
/// directory, so the tests work regardless of where CMake puts the binary.
std::filesystem::path GoldenVectorPath() {
  std::filesystem::path current = std::filesystem::current_path();
  for (int depth = 0; depth < 8; ++depth) {
    const std::filesystem::path candidate =
        current / "schemas" / "testdata" / "golden_vectors.json";
    if (std::filesystem::exists(candidate)) {
      return candidate;
    }
    if (!current.has_parent_path() || current.parent_path() == current) {
      break;
    }
    current = current.parent_path();
  }
  return {};
}

const nlohmann::json& GoldenVectors() {
  static const nlohmann::json data = [] {
    const std::filesystem::path path = GoldenVectorPath();
    REQUIRE_FALSE(path.empty());
    std::ifstream stream(path);
    REQUIRE(stream.is_open());
    nlohmann::json parsed;
    stream >> parsed;
    return parsed;
  }();
  return data;
}

std::optional<double> OptionalDouble(const nlohmann::json& value) {
  if (value.is_null()) {
    return std::nullopt;
  }
  return value.get<double>();
}

augury::DepthSide ParseSide(const std::string& name) {
  if (name == "bid") return augury::DepthSide::kBid;
  if (name == "ask") return augury::DepthSide::kAsk;
  return augury::DepthSide::kMean;
}

}  // namespace

// ---------------------------------------------------------------------------
// Golden vectors — cross-language agreement with the Python reference
// ---------------------------------------------------------------------------

TEST_CASE("golden vectors: cost and prices", "[golden][lmsr]") {
  for (const auto& entry : GoldenVectors()["lmsr"]["cost"]) {
    const auto q = entry["q"].get<std::vector<double>>();
    const double b = entry["b"].get<double>();

    CHECK(Cost(q, b) == Approx(entry["expected_cost"].get<double>()).margin(kTolerance));

    const auto expected_prices = entry["expected_prices"].get<std::vector<double>>();
    const std::vector<double> got = augury::Prices(q, b);
    REQUIRE(got.size() == expected_prices.size());
    for (std::size_t i = 0; i < got.size(); ++i) {
      CHECK(got[i] == Approx(expected_prices[i]).margin(kTolerance));
    }
  }
}

TEST_CASE("golden vectors: trade cost", "[golden][lmsr]") {
  for (const auto& entry : GoldenVectors()["lmsr"]["trade"]) {
    const auto q = entry["q"].get<std::vector<double>>();
    const auto dq = entry["dq"].get<std::vector<double>>();
    const double b = entry["b"].get<double>();
    CHECK(augury::TradeCost(q, dq, b) ==
          Approx(entry["expected_cost"].get<double>()).margin(kTolerance));
  }
}

TEST_CASE("golden vectors: binary price", "[golden][lmsr]") {
  for (const auto& entry : GoldenVectors()["lmsr"]["binary_price"]) {
    CHECK(PriceYes(entry["q_yes"].get<double>(), entry["q_no"].get<double>(),
                   entry["b"].get<double>()) ==
          Approx(entry["expected_price_yes"].get<double>()).margin(kTolerance));
  }
}

TEST_CASE("golden vectors: shares to move price", "[golden][lmsr]") {
  for (const auto& entry : GoldenVectors()["lmsr"]["shares_to_move"]) {
    CHECK(augury::SharesToMovePrice(entry["p_from"].get<double>(), entry["p_to"].get<double>(),
                                    entry["b"].get<double>()) ==
          Approx(entry["expected_shares"].get<double>()).margin(kTolerance));
  }
}

TEST_CASE("golden vectors: b calibration from book depth", "[golden][lmsr]") {
  for (const auto& entry : GoldenVectors()["lmsr"]["b_calibration"]) {
    const double got = augury::CalibrateBFromDepth(
        OptionalDouble(entry["depth_bid"]), OptionalDouble(entry["depth_ask"]),
        OptionalDouble(entry["yes_bid"]), OptionalDouble(entry["yes_ask"]),
        ParseSide(entry["side"].get<std::string>()), 100.0);
    CHECK(got == Approx(entry["expected_b"].get<double>()).margin(kTolerance));
  }
}

// ---------------------------------------------------------------------------
// Invariants — the same properties asserted by the Python suite
// ---------------------------------------------------------------------------

TEST_CASE("cost of a uniform state is b*ln(K)", "[lmsr]") {
  CHECK(Cost(std::vector<double>{0.0, 0.0}, 100.0) == Approx(100.0 * std::log(2.0)));
  CHECK(Cost(std::vector<double>{0.0, 0.0, 0.0, 0.0}, 50.0) == Approx(50.0 * std::log(4.0)));
}

TEST_CASE("cost does not overflow at realistic scale", "[lmsr]") {
  // q/b = 500. The naive formula evaluates exp(500), which is inf in IEEE
  // double, and every price derived from it becomes NaN.
  const double value = Cost(std::vector<double>{50000.0, 0.0}, 100.0);
  REQUIRE(std::isfinite(value));
  CHECK(value == Approx(50000.0).epsilon(1e-9));
}

TEST_CASE("cost is translation invariant", "[lmsr]") {
  const double base = Cost(std::vector<double>{10.0, -5.0}, 25.0);
  CHECK(Cost(std::vector<double>{110.0, 95.0}, 25.0) == Approx(base + 100.0));
}

TEST_CASE("prices form a probability distribution", "[lmsr]") {
  const std::vector<double> p = augury::Prices(std::vector<double>{120.0, -30.0, 45.0}, 40.0);
  double total = 0.0;
  for (double value : p) {
    CHECK(value > 0.0);
    CHECK(value < 1.0);
    total += value;
  }
  CHECK(total == Approx(1.0));
}

TEST_CASE("logit is linear in share imbalance", "[lmsr]") {
  const double b = 80.0;
  for (const auto& [q_yes, q_no] : std::vector<std::pair<double, double>>{
           {0.0, 0.0}, {100.0, 20.0}, {-60.0, 15.0}}) {
    const double p = PriceYes(q_yes, q_no, b);
    CHECK(std::log(p / (1.0 - p)) == Approx((q_yes - q_no) / b).margin(1e-9));
  }
}

TEST_CASE("a round trip is free up to float error", "[lmsr]") {
  LmsrMarket market(100.0);
  const double paid = market.BuyYes(75.0);
  const double refund = market.BuyYes(-75.0);
  CHECK(paid + refund == Approx(0.0).margin(1e-9));
  CHECK(market.Price() == Approx(0.5));
}

TEST_CASE("a YES share never costs more than 1", "[lmsr]") {
  // It pays at most 1 on resolution, so paying more would be arbitrage.
  LmsrMarket market(100.0);
  const double shares = 250.0;
  CHECK(market.BuyYes(shares) < shares);
}

TEST_CASE("moving to a target price is exact", "[lmsr]") {
  LmsrMarket market(60.0, 30.0, 0.0);
  for (double target : {0.1, 0.5, 0.97}) {
    market.MoveToPrice(target);
    CHECK(market.Price() == Approx(target).margin(1e-9));
  }
}

TEST_CASE("recalibrating b preserves the price", "[lmsr]") {
  // A liquidity update is not new information about the outcome; if it moved
  // the price it would inject signal nobody traded on.
  LmsrMarket market(100.0);
  market.MoveToPrice(0.73);
  const double before = market.Price();
  market.Recalibrate(925.0);
  CHECK(market.Price() == Approx(before).margin(1e-12));
  CHECK(market.b() == Approx(925.0));
}

TEST_CASE("higher b means less price impact", "[lmsr]") {
  LmsrMarket thin(10.0);
  LmsrMarket thick(10000.0);
  thin.BuyYes(100.0);
  thick.BuyYes(100.0);
  CHECK(std::abs(thin.Price() - 0.5) > std::abs(thick.Price() - 0.5));
}

TEST_CASE("calibrated b reproduces the observed price move", "[lmsr]") {
  // The whole justification for the formula: consuming `depth` shares should
  // walk the price from bid to ask.
  const double bid = 0.20;
  const double ask = 0.26;
  const double depth = 800.0;
  const double b = augury::CalibrateBFromDepth(depth, depth, bid, ask);

  LmsrMarket market(b);
  market.MoveToPrice(bid);
  market.BuyYes(depth);
  CHECK(market.Price() == Approx(ask).margin(1e-9));
}

TEST_CASE("unusable books fall back rather than throwing", "[lmsr]") {
  // Thin prediction markets produce these constantly.
  CHECK(augury::CalibrateBFromDepth(100.0, 100.0, 0.6, 0.4, augury::DepthSide::kMean, 42.0) == 42.0);
  CHECK(augury::CalibrateBFromDepth(100.0, 100.0, 0.5, 0.5, augury::DepthSide::kMean, 42.0) == 42.0);
  CHECK(augury::CalibrateBFromDepth(std::nullopt, std::nullopt, 0.3, 0.4,
                                    augury::DepthSide::kMean, 42.0) == 42.0);
  CHECK(augury::CalibrateBFromDepth(100.0, 100.0, std::nullopt, 0.4, augury::DepthSide::kMean,
                                    42.0) == 42.0);
}

TEST_CASE("invalid inputs are rejected", "[lmsr]") {
  CHECK_THROWS_AS(Cost(std::vector<double>{1.0}, 10.0), std::invalid_argument);
  CHECK_THROWS_AS(Cost(std::vector<double>{1.0, 2.0}, 0.0), std::invalid_argument);
  CHECK_THROWS_AS(LmsrMarket(100.0).StepFromSignal(0.6, 1.5), std::invalid_argument);
  CHECK_THROWS_AS(LmsrMarket(100.0).Recalibrate(0.0), std::invalid_argument);
}

TEST_CASE("signal steps converge on the target", "[lmsr]") {
  LmsrMarket market(100.0);
  for (int i = 0; i < 100; ++i) {
    market.StepFromSignal(0.7, 0.3);
  }
  CHECK(market.Price() == Approx(0.7).margin(1e-6));
}

TEST_CASE("max_shares caps a single step", "[lmsr]") {
  LmsrMarket market(100.0);
  const double shares = market.StepFromSignal(0.999, 1.0, 10.0);
  CHECK(shares == Approx(10.0));
}

// ---------------------------------------------------------------------------
// Backtester
// ---------------------------------------------------------------------------

namespace {

std::vector<augury::Observation> SyntheticSeries(std::size_t n) {
  std::vector<augury::Observation> series;
  series.reserve(n);
  double price = 0.5;
  for (std::size_t i = 0; i < n; ++i) {
    // Price follows a slow sine; the signal leads it by a few steps.
    price = 0.5 + 0.2 * std::sin(static_cast<double>(i) / 12.0);
    augury::Observation obs;
    obs.timestamp = static_cast<std::int64_t>(1'785'000'000 + i * 3600);
    obs.signal = std::sin((static_cast<double>(i) + 3.0) / 12.0);
    obs.market_price = price;
    obs.yes_bid = price - 0.01;
    obs.yes_ask = price + 0.01;
    obs.depth_bid = 500.0;
    obs.depth_ask = 500.0;
    series.push_back(obs);
  }
  return series;
}

}  // namespace

TEST_CASE("walk-forward produces multiple windows", "[backtest]") {
  const auto series = SyntheticSeries(300);
  augury::BacktestConfig config;
  config.train_size = 72;
  config.test_size = 24;

  const augury::BacktestSummary summary = augury::WalkForward(series, config);
  REQUIRE(summary.windows.size() > 1);
  CHECK(summary.b_min > 0.0);
  CHECK(summary.b_max >= summary.b_min);
  CHECK_FALSE(summary.Report().empty());
}

TEST_CASE("walk-forward test windows do not overlap by default", "[backtest]") {
  const auto series = SyntheticSeries(300);
  const augury::BacktestSummary summary = augury::WalkForward(series);
  for (std::size_t i = 1; i < summary.windows.size(); ++i) {
    CHECK(summary.windows[i].test_start >= summary.windows[i - 1].test_end);
  }
}

TEST_CASE("too little history yields no windows rather than a fake one", "[backtest]") {
  const auto series = SyntheticSeries(10);
  const augury::BacktestSummary summary = augury::WalkForward(series);
  CHECK(summary.windows.empty());
  CHECK(summary.Report().find("not enough") != std::string::npos);
}

TEST_CASE("mean squared error basics", "[backtest]") {
  CHECK(augury::MeanSquaredError({0.5, 0.5}, {0.5, 0.5}) == Approx(0.0));
  CHECK(augury::MeanSquaredError({1.0, 0.0}, {0.0, 1.0}) == Approx(1.0));
  CHECK(std::isnan(augury::MeanSquaredError({0.5}, {0.5, 0.5})));
}

TEST_CASE("difference correlation is undefined for constant series", "[backtest]") {
  CHECK(augury::DifferenceCorrelation({0.5, 0.5, 0.5, 0.5}, {0.1, 0.2, 0.3, 0.4}) == Approx(0.0));
}

TEST_CASE("difference correlation detects a perfect match", "[backtest]") {
  const std::vector<double> series{0.1, 0.2, 0.35, 0.5, 0.62, 0.7};
  CHECK(augury::DifferenceCorrelation(series, series) == Approx(1.0).margin(1e-9));
}
